### PR TITLE
Trace key operations

### PR DESCRIPTION
**Pull request by amcasey**
Produce output in Chrome's event tracing format so that it can be viewed in Chrome/Edge.

Hat tip to @samccone for suggesting the format.

### CODE REVIEW

**Comment by amcasey** [comment]
There are some noteworthy rough edges:
1) I haven't thought through how to separate/combine the traces for different projects in build mode, so presently I just insert a counter into each path to prevent clobbering.
2) The new System APIs are ugly, but I think it's important to write as you go, so that you get something if you have to kill a very long build (e.g. customers reporting tens of minutes)
3) Some types are still only readable by means of `typeToTypeNode` and I'd really like to eliminate that because of how slow it is (not urgent since it happens after profiling is complete)

**Comment by amcasey** [comment]
CI failures are because I didn't update the harness System impl(s).  That can wait until next week.

**Comment by amcasey** [comment]
There are deliberately few trace events in the checker to avoid (excessively) skewing the runtime.  If people would like to see more, e.g. for debugging, my inclination would be to add a separate mode (i.e. profiling trace vs activity trace).  Note that the persistence format is JSON, so it's very easy to create enormous trace files (e.g. adding `inferFromTypes` tripled the file size for one sample project).

**Comment by amcasey** [comment]
Having applied this to several repos now, I feel pretty confident that this is useful in its current state.  Enriching it with additional tracepoints (e.g. for inference) or output (e.g. stack traces) can follow in future PRs.

File: src/compiler/tracing.ts [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+    let traceFd: number | undefined;
+
+    let legendPath: string | undefined;
+    const legend: TraceRecord[] = [];
+
+    /** Starts tracing for the given project (unless the `fs` module is unavailable). */
+    export function startTracing(configFilePath: string | undefined, traceDir: string, isBuildMode: boolean) {
+        Debug.assert(!traceFd, "Tracing already started");
+
+        if (fs === undefined) {
```
Comment (rbuckton) [inline src/compiler/tracing.ts:16]: In the future we might be able to move some of this logic to a conditionally loaded module, similar to the extended debug information for flow nodes in _~/src/debug/dbg.ts_ loaded by _~/src/compiler/debug.ts_: https://github.com/microsoft/TypeScript/blob/d89635381a65229c5ab5b0c73de15347b86235a2/src/compiler/debug.ts#L655

I've been thinking about moving more of `ts.Debug` into the external file to reduce the file size of _tsc.js_, _typescript.js_, and _tsserver.js_ as well.

This obviously isn't something we need to do now, but we might want to consider putting it on the backlog.

**Comment by phryneas** [comment]
@amcasey The problem is that `checkExpression` calls within a `checkSourceFile` call often check nodes from completely different files, so just having `kind`, `pos` and `end` is not enough.

I am running this patch against 4.1.0-beta to also get the `fileName` - but I guess that is not really effective.
Could this be done somehow in a better way - and would a PR for this be welcome?
```diff
diff --git a/lib/tsc.js b/lib/tsc.js
index c78b8d698ea730be5a363592a25eb5eb9815753b..4333d58e45d2b2a55724ce7f7894c482223fc36c 100644
--- a/lib/tsc.js
+++ b/lib/tsc.js
@@ -59317,7 +59317,13 @@ var ts;
             }
         }
         function checkExpression(node, checkMode, forceTuple) {
-            ts.tracing.begin("check", "checkExpression", { kind: node.kind, pos: node.pos, end: node.end });
+            var parent = node.parent;
+            while (parent && parent.kind !== 300) {
+                parent = parent.parent
+            }
+            var sourceFile = parent;
+
+            ts.tracing.begin("check", "checkExpression", { kind: node.kind, pos: node.pos, end: node.end, sourceFile: sourceFile && sourceFile.fileName || "unknown" });
             var saveCurrentNode = currentNode;
             currentNode = node;
             instantiationCount = 0;

```

**Comment by amcasey** [comment]
@phryneas Thanks for the patch!  I had an early version that included similar code (there's actually a helper for walking up to the source file) but concluded that the extra tree walking was distorting the timing and the paths were bloating the log file excessively.  Since readability wasn't a goal for v1, I left it out.  When I've encountered the problem in practice, I've used code inspection on the frames that _are_ in the right file plus types from later in the call stack to work out what was going on.

p.s.  Sorry about the confusion this caused you - it's frustrating to waste time on things like that.

**Pull request merged**